notation/notation.go

```diff
@@ -0,0 +1,314 @@
+	TrustPolicyDocument() (*trustpolicy.Document, error)
+}
+
+/*
+Verify performs signature verification on each of the notation supported
+verification types (like integrity, authenticity, etc.) and return the
+verification outcomes.
+
+Given an artifact reference, Verify will retrieve all the signatures associated
+with the reference and perform signature verification.
+A signature is considered not valid if verification fails due to any one of the
+following reasons
+
+ 1. Artifact Reference is not associated with a signature i.e. unsigned
+ 2. Registry is unavailable to retrieve the signature
+ 3. Signature does not satisfy the verification rules configured in the trust
+    policy
+ 4. Signature specifies a plugin for extended verification and that throws an
+    error
+ 5. Digest in the signature does not match the digest present in the reference
+
+If each and every signature associated with the reference fail the verification,
+then Verify will return `ErrorVerificationFailed` error along with an array
+of `VerificationOutcome`.
+
+# Callers can pass the verification plugin config in VerifyOptions.PluginConfig
+
+For more details on signature verification, see https://github.com/notaryproject/notaryproject/blob/main/trust-store-trust-policy-specification.md#signature-verification
+*/
+func Verify(ctx context.Context, verifier Verifier, repo registry.Repository, opts VerifyOptions) (Descriptor, []*VerificationOutcome, error) {
+	var verificationOutcomes []*VerificationOutcome
+	artifactRef := opts.ArtifactReference
+	artifactDescriptor, err := repo.Resolve(ctx, artifactRef)
+	if err != nil {
+		return Descriptor{}, nil, ErrorSignatureRetrievalFailed{Msg: err.Error()}
+	}
+
+	trustpolicyDoc, err := verifier.TrustPolicyDocument()
+	if err != nil {
+		return Descriptor{}, nil, ErrorNoApplicableTrustPolicy{Msg: err.Error()}
+	}
+	trustPolicy, err := trustpolicy.GetApplicableTrustPolicy(trustpolicyDoc, artifactRef)
+	if err != nil {
+		return Descriptor{}, nil, ErrorNoApplicableTrustPolicy{Msg: err.Error()}
+	}
+	// ignore the error since we already validated the policy document
+	verificationLevel, _ := trustpolicy.GetVerificationLevel(trustPolicy.SignatureVerification)
+	if verificationLevel.Name == trustpolicy.LevelSkip.Name {
+		verificationOutcomes = append(verificationOutcomes, &VerificationOutcome{VerificationLevel: verificationLevel})
+		return Descriptor{}, verificationOutcomes, nil
+	}
+	opts.VerificationLevel = verificationLevel
+
+	// get signature manifests
+	var success bool
+	var verifiedSigBlobDesc ocispec.Descriptor
+	err = repo.ListSignatures(ctx, artifactDescriptor, func(signatureManifests []ocispec.Descriptor) error {
+		if len(signatureManifests) < 1 {
+			return ErrorSignatureRetrievalFailed{Msg: fmt.Sprintf("no signatures are associated with %q, make sure the image was signed successfully", artifactRef)}
+		}
+		// if already verified successfully, no need to continue
+		if success {
+			return nil
+		}
```

Comment:
If success is set to true do we need to perform `len(signatureManifests) < 1` check?

Reply:
Nice catch! Moved `len(signatureManifests) < 1` after checking `success`.